### PR TITLE
Require authorization on DAR upload endpoint

### PR DESCRIPTION
**Pull request by aherrmann-da**
Requires authorization for `admin` claims for the upload DAR endpoint of the trigger service.

The existing `*Auth` test-suites test that uploading a DAR succeeds when permission is granted. This PR adds a test-case for the case where uploading a DAR is forbidden.

An upload DAR request contains a DAR in the HTTP entity. By default this upload is streamed. However, when a redirect for a login flow is required we need to make sure that the entity is fully uploaded before we reply with a redirection. Otherwise, the continuation will be left with a truncated entity. To that end we use the `toStrictEntity` directive. This PR adds corresponding CLI flags for the max upload size and upload timeout.

Part of #7723

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala [hunk 1]
```diff
@@ -134,6 +138,22 @@ private[trigger] object ServiceConfig {
       .text(
         s"Maximum time interval between restarting a failed trigger. Defaults to ${DefaultMaxRestartInterval.toSeconds} seconds.")
 
+    opt[Int]("max-http-entity-upload-size")
+      .action((x, c) => c.copy(maxHttpEntityUploadSize = x))
+      .optional()
+      .text(
+        s"Optional max HTTP entity upload size. Defaults to ${DefaultMaxHttpEntityUploadSize}. HTTP uploads cannot be streamed but must be fully uploaded when a request is deferred until after a login flow with the auth middleware completed.")
+      // TODO[AH] Expose once the auth feature is fully implemented.
+      .hidden()
```
Comment (aherrmann-da) [inline triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala:147]: The auth feature is not yet exposed (`.hidden`) so it doesn't make sense to expose these flags independently.

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala [hunk 2]
```diff
@@ -134,6 +138,22 @@ private[trigger] object ServiceConfig {
...
+
+    opt[Long]("http-entity-upload-timeout")
+      .action((x, c) => c.copy(httpEntityUploadTimeout = FiniteDuration(x, duration.MINUTES)))
+      .optional()
+      .text(
+        s"Optional HTTP entity upload timeout. Defaults to ${DefaultHttpEntityUploadTimeout.toSeconds} seconds. HTTP uploads cannot be streamed but must be fully uploaded when a request is deferred until after a login flow with the auth middleware completed.")
```
Comment (cocreature) [inline triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala:153]: The streaming detail might be a bit much for `--help` output.

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala [hunk 3]
```diff
@@ -81,6 +83,8 @@ private[trigger] object ServiceConfig {
   val DefaultMaxInboundMessageSize: Int = RunnerConfig.DefaultMaxInboundMessageSize
   private val DefaultMinRestartInterval: FiniteDuration = FiniteDuration(5, duration.SECONDS)
   val DefaultMaxRestartInterval: FiniteDuration = FiniteDuration(60, duration.SECONDS)
+  val DefaultMaxHttpEntityUploadSize: Int = 10485760
```
Comment (cocreature) [inline triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala:86]: What is this based on?

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Server.scala [hunk 4]
```diff
@@ -268,40 +270,43 @@ class Server(
...
-                      // Authorization successful after login - use old request context and pass token to continuation.
-                      mapRequestContext(_ => ctx) {
-                        inner(Tuple1(Some(authorization)))
+        auth.flatMap {
+          // Authorization successful - pass token to continuation
+          case Some(authorization) => provide(Some(authorization))
+          // Authorization failed - login and retry on callback request.
+          case None =>
+            // Ensure that the request is fully uploaded.
+            toStrictEntity(httpEntityUploadTimeout, maxHttpEntityUploadSize.toLong).tflatMap { _ =>
```
Comment (cocreature) [inline triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Server.scala:279]: Maybe we should just store it as a long given that we only ever use it as that?

**Review by cocreature — APPROVED** [review]
Thank you!

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala (hunk 2 above)
Comment (aherrmann-da) [reply triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala:153]: Maybe so, I've dropped it.

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Server.scala (hunk 4 above)
Comment (aherrmann-da) [reply triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Server.scala:279]: Good point, fixed.

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala (hunk 3 above)
Comment (aherrmann-da) [reply triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala]: Based on our discussion on Slack on what should be big enough for most applications. I've set it to `RunnerConfig.DefaultMaxInboundMessageSize` to not have too many such numbers lying around.

**Pull request merged**